pkg/cloud/ratelimit.go

```diff
@@ -83,6 +85,49 @@ func (rl *AcceptRateLimiter) Accept(ctx context.Context, _ *RateLimitKey) error
 func (rl *AcceptRateLimiter) Observe(context.Context, error, *RateLimitKey) {
 }
 
+// ThrottlingStrategy handles delays based on feedbacks provided.
+type ThrottlingStrategy interface {
+	// Delay returns the delay for next request.
+	Delay() time.Duration
+	// Observe recalculates the next delay based on the feedback.
+	Observe(err error)
+}
+
+// StrategyRateLimiter wraps a ThrottlingStrategy with RateLimiter parameters.
+type StrategyRateLimiter struct {
+	lock sync.Mutex
+	// strategy is the underlying throttling strategy.
+	strategy ThrottlingStrategy
+	
+	clock clock.Clock
+}
+
+// NewStrategyRateLimiter returns a StrategyRateLimiter backed by the provided ThrottlingStrategy.
+func NewStrategyRateLimiter(strategy ThrottlingStrategy) *StrategyRateLimiter {
+	return &StrategyRateLimiter{
+		strategy: strategy,
+		clock:    clock.RealClock{},
+	}
+}
+
+// Accept block for the delay provided by the strategy or until context.Done(). Key is ignored.
+func (rl *StrategyRateLimiter) Accept(ctx context.Context, _ *RateLimitKey) error {
+	rl.lock.Lock()
```

Comment:
It's not clear to me why this lock is being used. What is it guarding?

Reply:
It's hard to tell about what's guarded exactly, it's kind of guarding a combination of `Delay` call to strategy and sleeping. Here's a motivation for it:
The `Delay()` function of a strategy returns the amount of time to wait before sending the request. But the strategy isn't keeping track of requests, so the delay for 2 requests coming at the same time will be the same, meaning we won't get any delay between first and second requests. With the lock we'll wait until the first request is allowed to go, and then the second request can get its delay value.

There's a possibility for a strategy to return not the delay, but the timestamp to wait until (keeping more info inside each strategy implementation), but it will make more time for a recalculated delay to come into effect. Here's an example of a situation with and without the lock:

**Without the lock**:
Current delay for a strategy is 5 seconds.
1. Requests 1-3 come at T=0 and get timestamps for proceeding with the request T=5, T=10, T=15
2. Request 1 starts the execution at T=5 and got the response at T=6, which is observed by the strategy leading to setting the delay to 1 second.
3. Requests 2-3 won't get updated timestamp for proceeding, so the last request will still start at T=15.

**With the lock**:
Current delay for a strategy is 5 seconds.
1. Requests 1-3 come at T=0, request 1 gets delay of 5 seconds with the lock, while requests 2-3 are blocked.
2. Request 1 starts the execution at T=5, request 2 gets delay of 5 seconds (until T=10) with the lock.
3. Request 1 got the response at T=6, which is observed by the strategy leading to setting the delay to 1 second.
4. Request 2 starts the execution at T=10, request 3 gets delay of 1 second (until T=11) with the lock.
5. Request 3 starts the execution at T=11.

Reply:
I think this might need to consider the RateLimitKey, no? With this implementation, you will end up queueing all of the requests, not just ones with a given API. I think you need to maintain queues for each API call?

Reply:
In general yes, but the idea behind was to wrap `StrategyRateLimiter` with some other `RateLimiter` to consider `RateLimitKey` there, similar to `AcceptRateLimiter`. It makes sense though to add an explicit comment about the usage of such a rate limiter. Wdyt?